### PR TITLE
[CARBONDATA-2078][CARBONDATA-1516] Add 'if not exists' function for creating datamap

### PR DESCRIPTION
**Pull request by xubo245**
Be sure to do all of the following checklist to help us incorporate 
your contribution quickly and easily:

 - [x] Any interfaces changed?
Add 'if not exists' function for creating datamap

```

   The syntax of datamap creation is as follows.
   CREATE DATAMAP IF NOT EXISTS datamapName ON TABLE tableName USING 'DataMapClassName'
    DMPROPERTIES('KEY'='VALUE') AS SELECT COUNT(COL1) FROM tableName
```

 - [x] Any backward compatibility impacted?
 No
 - [x] Document update required?
No
 - [x] Testing done
 add some test case in org.apache.carbondata.integration.spark.testsuite.preaggregate.TestPreAggCreateCommand

and org.apache.carbondata.integration.spark.testsuite.timeseries.TestTimeSeriesCreateTable
 - [x] For large changes, please consider breaking it into sub-tasks under an umbrella JIRA. 

refer CARBONDATA-1516

### CODE REVIEW

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/3130/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.2.1, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/1897/

**Comment by BJangir** [comment]
@xubo245 , 
 if agg table is already created and create datamap command is fired again with "if not exists" syntax then processMetadata will not return any Error (as expected ) but  then processData will be triggered
 and processData will throw error because createPreAggregateTableCommands is not initialized . So "table exists"  check need to handle in processData also .  Steps to reproduce 
1. Create table maintable 
2. load data maintable 
3. Create datamap map1 on table maintable
4. Create datamap if not  exists map1 on table maintable

**Comment by ravipesala** [comment]
SDV Build Success , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/3107/

**Comment by xubo245** [comment]
createPreAggregateTableCommands will be none  in processData when we create datamap with "if not exists" syntax .

so When will throw error? can you give a example for " processData will throw error because createPreAggregateTableCommands is not initialized"?

**Comment by xubo245** [comment]
@BJangir Can you give me the code of throw error?

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/3141/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.2.1, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/1908/

**Comment by ravipesala** [comment]
SDV Build Success , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/3115/

**Comment by kumarvishal09** [comment]
@xubo245 Please use the step provided by @BJangir and try to reproduce the issue.

**Comment by xubo245** [comment]
@kumarvishal09 @BJangir I tried before, but not happen according the step, please see: https://github.com/apache/carbondata/pull/1861/commits/723757f3343baa7d47d532ccdda268597023b769

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/3196/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.2.1, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/1962/

**Comment by ravipesala** [comment]
SDV Build Success , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/3159/

**Comment by xubo245** [comment]
@chenliang613 @jackylk @BJangir @kumarvishal09 @sraghunandan Please review it.

**Comment by kumarvishal09** [comment]
@xubo245 Please rebase

**Comment by kumarvishal09** [comment]
@xubo245 After merging #1865 we need to update the createdatamap class as now user will able only one timeseries table in one create data map command please change the code accordingly

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.2.1, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/2039/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/3278/

**Comment by xubo245** [comment]
@kumarvishal09 done, please review https://github.com/apache/carbondata/pull/1861/commits/6537d1606282db8418ac580ce476c7ea8a856a0c

**Comment by xubo245** [comment]
Retest this please

**Comment by ravipesala** [comment]
SDV Build Success , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/3230/

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/3301/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.2.1, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/2064/

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/3306/

**Comment by CarbonDataQA** [comment]
Build Failed with Spark 2.2.1, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/2069/

**Comment by ravipesala** [comment]
SDV Build Success , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/3239/

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/3333/

**Comment by CarbonDataQA** [comment]
Build Failed with Spark 2.2.1, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/2097/

File: integration/spark2/src/main/scala/org/apache/spark/sql/execution/command/datamap/CarbonCreateDataMapCommand.scala [hunk 1]
```diff
@@ -49,10 +61,21 @@ case class CarbonCreateDataMapCommand(
       throw new MalformedCarbonCommandException("Streaming table does not support creating datamap")
     }
     val LOGGER = LogServiceFactory.getLogService(this.getClass.getCanonicalName)
+    val dbName = tableIdentifier.database.getOrElse("default")
+    val tableName = tableIdentifier.table + "_" + dataMapName
 
-    if (dmClassName.equalsIgnoreCase(PREAGGREGATE.toString) ||
+    if (tableExists(sparkSession, dbName, tableName)) {
```
Comment (kumarvishal09) [inline integration/spark2/src/main/scala/org/apache/spark/sql/execution/command/datamap/CarbonCreateDataMapCommand.scala]: tableExists method is called only from processMeta, so better not to create any method you can update the if statement
if(sparkSession.sessionState.catalog.listTables(dbName) .exists(_.table.equalsIgnoreCase(tableName)))

Comment (xubo245) [reply integration/spark2/src/main/scala/org/apache/spark/sql/execution/command/datamap/CarbonCreateDataMapCommand.scala]: ok,done

**Comment by CarbonDataQA** [comment]
Build Failed  with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/3345/

**Comment by CarbonDataQA** [comment]
Build Failed with Spark 2.2.1, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/2109/

**Comment by ravipesala** [comment]
SDV Build Success , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/3257/

**Comment by ravipesala** [comment]
SDV Build Success , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/3270/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/3382/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.2.1, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/2145/

**Comment by xubo245** [comment]
CI pass @kumarvishal09

File: integration/spark-common-test/src/test/scala/org/apache/carbondata/integration/spark/testsuite/timeseries/TestTimeSeriesCreateTable.scala [hunk 2]
```diff
@@ -81,29 +82,29 @@ class TestTimeSeriesCreateTable extends QueryTest with BeforeAndAfterAll {
        """.stripMargin)
   }
 
-  test("test timeseries create table Zero") {
+  test("test timeseries create table 1") {
```
Comment (kunal642) [inline integration/spark-common-test/src/test/scala/org/apache/carbondata/integration/spark/testsuite/timeseries/TestTimeSeriesCreateTable.scala]: Please remove unnecessary changes like this

File: integration/spark-common-test/src/test/scala/org/apache/carbondata/integration/spark/testsuite/preaggregate/TestPreAggregateLoad.scala [hunk 3]
```diff
@@ -310,5 +310,99 @@ test("check load and select for avg double datatype") {
     checkAnswer(sql("select name,avg(salary) from maintbl group by name"), rows)
   }
 
+  test("create datamap with 'if not exists' after load data into mainTable and create datamap") {
```
Comment (kunal642) [inline integration/spark-common-test/src/test/scala/org/apache/carbondata/integration/spark/testsuite/preaggregate/TestPreAggregateLoad.scala]: I think no need to add test cases in all the files. One test case in TestPreAggregateCreateCommand and one in TestTimeseriesCreateTable would be enough.

File: integration/spark-common-test/src/test/scala/org/apache/carbondata/integration/spark/testsuite/timeseries/TestTimeSeriesCreateTable.scala [hunk 4]
```diff
@@ -319,6 +326,53 @@ class TestTimeSeriesCreateTable extends QueryTest with BeforeAndAfterAll {
     assert(e.getMessage.equals(s"$timeSeries should define time granularity"))
   }
 
+  test("test timeseries create table 19: should support if not exists") {
+    sql("DROP DATAMAP IF EXISTS agg1 ON TABLE mainTable")
+    try {
```
Comment (kunal642) [inline integration/spark-common-test/src/test/scala/org/apache/carbondata/integration/spark/testsuite/timeseries/TestTimeSeriesCreateTable.scala]: no need for try block. If any exception if thrown the test case will fail

File: integration/spark-common-test/src/test/scala/org/apache/carbondata/integration/spark/testsuite/timeseries/TestTimeSeriesCreateTable.scala (hunk 2 above)
Comment (xubo245) [reply integration/spark-common-test/src/test/scala/org/apache/carbondata/integration/spark/testsuite/timeseries/TestTimeSeriesCreateTable.scala]: It's necessary and convenience to find the test case according to the number when some test case failed.

File: integration/spark-common-test/src/test/scala/org/apache/carbondata/integration/spark/testsuite/preaggregate/TestPreAggregateLoad.scala (hunk 3 above)
Comment (xubo245) [reply integration/spark-common-test/src/test/scala/org/apache/carbondata/integration/spark/testsuite/preaggregate/TestPreAggregateLoad.scala]: not all file, but I think we should add this test case:  Create data map when the main table has data.

Comment (xubo245) [reply integration/spark-common-test/src/test/scala/org/apache/carbondata/integration/spark/testsuite/preaggregate/TestPreAggregateLoad.scala]: And another problem happend when the order of the load data into main table and create datamap is different, you can check CARBONDATA-2085.

File: integration/spark2/src/main/scala/org/apache/spark/sql/execution/command/datamap/CarbonCreateDataMapCommand.scala [hunk 5]
```diff
@@ -49,10 +52,22 @@ case class CarbonCreateDataMapCommand(
       throw new MalformedCarbonCommandException("Streaming table does not support creating datamap")
     }
     val LOGGER = LogServiceFactory.getLogService(this.getClass.getCanonicalName)
+    val dbName = tableIdentifier.database.getOrElse("default")
+    val tableName = tableIdentifier.table + "_" + dataMapName
 
-    if (dmClassName.equalsIgnoreCase(PREAGGREGATE.toString) ||
+    if (sparkSession.sessionState.catalog.listTables(dbName)
```
[the remark this answers is not in the document]
Comment (xubo245) [reply integration/spark2/src/main/scala/org/apache/spark/sql/execution/command/datamap/CarbonCreateDataMapCommand.scala]: When I change it to "sparkSession.sessionState.catalog.tableExists(tableIdentifier)", it will fail. like:

An exception or error caused a run to abort: Table or view 'mainTable_agg0_second' already exists in database 'default'; 
org.apache.spark.sql.catalyst.analysis.TableAlreadyExistsException: Table or view 'mainTable_agg0_second' already exists in database 'default';

File: integration/spark-common-test/src/test/scala/org/apache/carbondata/integration/spark/testsuite/timeseries/TestTimeSeriesCreateTable.scala (hunk 4 above)
Comment (xubo245) [reply integration/spark-common-test/src/test/scala/org/apache/carbondata/integration/spark/testsuite/timeseries/TestTimeSeriesCreateTable.scala]: ok, done

**Comment by kunal642** [comment]
LGTM

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.2.1, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/2173/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/3415/

**Comment by ravipesala** [comment]
SDV Build Success , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/3286/

**Comment by ravipesala** [comment]
SDV Build Success , Please check CI http://144.76.159.231:8080/job/ApacheSDVTests/3294/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.1.0, Please check CI http://136.243.101.176:8080/job/ApacheCarbonPRBuilder1/3427/

**Comment by CarbonDataQA** [comment]
Build Success with Spark 2.2.1, Please check CI http://88.99.58.216:8080/job/ApacheCarbonPRBuilder/2188/
